client/lib/logmein/index.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * Internal Dependencies
+ */
+import { isEnabled } from '@automattic/calypso-config';
+import wpcom from 'calypso/lib/wp';
+
+/**
+ * Append logmein=1 query parameter to mapped domain urls we want the user to be logged in against.
+ */
+
+// Used as placeholder / default domain to detect when we're looking at a relative url
+const INVALID_URL = `https://__domain__.invalid`;
+
+export function logmeinUrl( url: string, sites: any, isWPComLoggedIn = true ): string {
+	let newurl: URL;
+
+	if ( ! isEnabled( 'logmein' ) ) {
+		return url;
+	}
+
+	try {
+		newurl = new URL( String( url ), INVALID_URL );
+	} catch ( e ) {
+		// Ignore unparseable urls
+		return url;
+	}
+
+	// Ignore and passthrough /relative/urls that have no host specified
+	if ( newurl.origin === INVALID_URL ) {
+		return url;
+	}
+
+	let allow = logmeinAllowedUrls( sites );
+	// Ignore urls not in the allow list
+	allow = allow.map( ( allowed: string ) => new URL( allowed, INVALID_URL ).hostname );
+	if ( allow.indexOf( newurl.hostname ) === -1 ) {
+		return url;
+	}
+
+	// logmein doesn't work with http.
+	newurl.protocol = 'https:';
+
+	// we're already logged into wordpress.com so we can take a shortcut
+	if ( isWPComLoggedIn ) {
+		newurl.searchParams.set( 'logmein', 'direct' );
+	} else {
+		newurl.searchParams.set( 'logmein', '1' );
+	}
+	return newurl.toString();
+}
+
+export function logmeinAllowedUrls( sites: any ): string[] {
+	return sites
+		.map( ( site: any ) =>
+			! site.is_vip &&
+			! site.jetpack &&
+			! site.options.is_automated_transfer &&
+			! site.options.is_domain_only &&
+			! site.options.is_redirect &&
+			! site.options.is_wpcom_atomic &&
+			! site.options.is_wpcom_store &&
+			! site.options.is_wpforteams_site &&
+			site.options.is_mapped_domain
+				? new URL( site.URL ).host
+				: false
+		)
+		.filter( Boolean );
+}
+
+export function appendLogmein( url: URL ): URL {
+	// logmein doesn't work with http.
+	url.protocol = 'https:';
+	url.searchParams.set( 'logmein', '1' );
+	return url;
+}
+export function appendLogmeinDirect( url: URL ): URL {
+	// logmein doesn't work with http.
+	url.protocol = 'https:';
+	url.searchParams.set( 'logmein', 'direct' );
+	return url;
+}
+
+let allowedSites: string[] = [];
+
+export function attachLogmein( isWPComLoggedIn = false ): void {
+	if ( ! isEnabled( 'logmein' ) ) {
+		return;
+	}
+
+	if ( isWPComLoggedIn ) {
+		wpcom
+			.me()
+			.sites()
+			.then( ( sites: any ) => {
+				allowedSites = logmeinAllowedUrls( sites.sites );
+				document.addEventListener( 'click', logmeinOnClick, false );
+			} );
+	}
+}
+
+function logmeinOnClick( event: MouseEvent ) {
+	const link = ( event.target as HTMLElement ).closest( 'a' );
+
+	if ( link ) {
+		let url = new URL( link.href, INVALID_URL );
+		if ( allowedSites.indexOf( url.hostname ) !== -1 ) {
+			url = appendLogmeinDirect( url );
+			console.log( 'intercepted', link.href, 'replaced', url.toString() );
+			link.href = url.toString();
```

Comment:
This changes the actual `href` attribute of the `a` element and the default browser handler will use that to navigate. That's cool and smart 🙂 In some edge cases it might do that repeatedly though, on each click.

Reply:
It probably will, though I think I used the right `set` call on URL to make that safe (will double check)

I was thinking I could clear the value after but its probably overkill.

Reply:
This is a novel approach, I hadn't thought there'd be a way of fixing all the links in one hit like this does.

But will this approach break the right-click to open in a new tab/window experience? Perhaps there's a way of getting a click handler to fire even when the link is right-clicked so we can adjust the `href` in that case too. But then there's probably a keyboard driven way to open in a new tab/window too. These are downsides that the `localizeUrl()` approach doesn't have (even though it'd mean peppering a call to `logmeinUrl()` all through the code.

Reply:
Right click open doesn't work with this PR as is, [should](https://stackoverflow.com/a/4236294) be able to detect right click events using [`contextmenu`](https://developer.mozilla.org/en-US/docs/Web/API/Element/contextmenu_event) events whether these happen at the right time needs checking.

Keyboard `tab` to focus + `enter` does.

Looking up options it [might be possible](https://stackoverflow.com/questions/3522090/event-when-window-location-href-changes) to track the href value before navigation using [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver).

It might also be worth checking if there is any undue performance hit using MutationObserver to watch and check links as they're added to the dom. I've done similar things in the past but it was expensive. I wasn't using mutation observers though so maybe it would work better?

Reply:
TIL about the `contextmenu` event 👍

Reply:
Added contextmenu event handler, seems to fire on right click so you don't need to perform any action in the context menu and the link still gets updated.

Reply:
> In some edge cases it might do that repeatedly though, on each click.
Also added a `seen[host]` check so we'll only set it once per click, right click always applies as you need to ensure the user visits the link before you know its safe to stop adding it.